src/promote/diff_promotion.ml

```diff
@@ -149,47 +147,61 @@ let do_promote db files_to_promote =
          not promote into the build directory anyway), and source digests should
          be correctly invalidated via [fs_memo]. If that doesn't happen, we
          should fix [fs_memo] instead of manually resetting the caches here. *)
-      File.promote { src; staging; dst };
-      List.iter others ~f:(fun (path, _staging) ->
-        Console.print
-          [ Pp.textf " -> ignored %s." (Path.to_string_maybe_quoted (Path.build path))
-          ; Pp.space
-          ])
+    File.promote { src; staging; dst };
+    List.iter others ~f:(fun (path, _staging) ->
+      Console.print
+        [ Pp.textf " -> ignored %s." (Path.to_string_maybe_quoted (Path.build path))
+        ; Pp.space
+        ])
+;;
+
+let do_promote_all db = group_by_targets db |> Path.Source.Map.iteri ~f:promote_one
+
+let do_promote_these db files =
+  let by_targets = group_by_targets db in
+  let by_targets, missing =
+    let files = Path.Source.Set.of_list files in
+    Path.Source.Set.fold files ~init:(by_targets, []) ~f:(fun fn (map, missing) ->
+      match Path.Source.Map.find map fn with
+      | None -> map, fn :: missing
+      | Some srcs ->
+        promote_one fn srcs;
+        Path.Source.Map.remove map fn, missing)
   in
-  match files_to_promote with
-  | Dune_rpc_private.Files_to_promote.All ->
-    Path.Source.Map.iteri by_targets ~f:promote_one;
-    []
-  | These (files, on_missing) ->
-    let by_targets =
-      let files = Path.Source.Set.of_list files in
-      Path.Source.Set.fold files ~init:by_targets ~f:(fun fn map ->
-        match Path.Source.Map.find by_targets fn with
-        | None ->
-          on_missing fn;
-          map
-        | Some srcs ->
-          promote_one fn srcs;
-          Path.Source.Map.remove by_targets fn)
-    in
+  let remaining =
     Path.Source.Map.to_list by_targets
     |> List.concat_map ~f:(fun (dst, srcs) ->
       List.map srcs ~f:(fun (src, staging) -> { File.src; staging; dst }))
+  in
+  let sorted_missing = List.rev missing in
```

Comment:
`group_by_targets` will sort all files, but the `fold` above reverses the list of missing files.
The absence of this reverse is what had lead to a swap in `test-cases/promote/old-tests.t` that was present in this PR before I noticed